src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java

```diff
@@ -52,57 +54,75 @@ public ContainerFactory(final SAMFileHeader samFileHeader, final int recordsPerS
-            slice.globalRecordCounter = lastGlobalRecordCounter;
-            lastGlobalRecordCounter += slice.nofRecords;
-            container.bases += slice.bases;
-            slices.add(slice);
-
-            // assuming one sequence per container max:
-            if (container.sequenceId == -1 && slice.sequenceId != -1)
-                container.sequenceId = slice.sequenceId;
+        // write to a stream to determine byte sizes for Slice BAI Metadata
+        try (final ByteArrayOutputStream osForIndexMetadata = new ByteArrayOutputStream()) {
+            final Version version = CramVersions.CRAM_v3;
+            container.header.write(version, osForIndexMetadata);
+
+            // TODO: is this copy necessary?
+            long globalRecordCounterCopy = container.globalRecordCounter;
+            int sliceIndex = 0;
+            for (int i = 0; i < records.size(); i += recordsPerSlice) {
+                final List<CramCompressionRecord> sliceRecords = records.subList(i,
+                        Math.min(records.size(), i + recordsPerSlice));
+                final Slice slice = buildSlice(sliceRecords, header, globalRecordCounterCopy, refBases);
+                globalRecordCounterCopy += sliceRecords.size();
+
+                // TODO for Container refactoring:
+                // these are also the container.landmarks so let's centralize their construction here
+
+                final int sliceByteOffset = osForIndexMetadata.size();
+                slice.write(version.major, osForIndexMetadata);
+                final int sliceByteSize = osForIndexMetadata.size() - sliceByteOffset;
+
+                slices.add(slice.withIndexingMetadata(sliceByteOffset, sliceByteSize, sliceIndex++));
+
+                container.bases += sliceRecords.stream().map(r -> r.readLength).reduce(Integer::sum).orElse(0);
+
+                // assuming one sequence per container max:
+                if (container.sequenceId == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX && !slice.hasNoReference())
+                    container.sequenceId = slice.getSequenceId();
+            }
+        }
+        catch (final IOException e) {
+            throw new RuntimeIOException(e);
         }
 
-        container.slices = slices.toArray(new Slice[slices.size()]);
+        container.slices = slices.toArray(new IndexableSlice[0]);
         calculateAlignmentBoundaries(container);
 
         globalRecordCounter += records.size();
+
         return container;
     }
 
     private static void calculateAlignmentBoundaries(final Container container) {
         int start = Integer.MAX_VALUE;
         int end = Integer.MIN_VALUE;
-        for (final Slice s : container.slices) {
-            if (s.sequenceId != SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX) {
-                start = Math.min(start, s.alignmentStart);
-                end = Math.max(end, s.alignmentStart + s.alignmentSpan);
+        for (final Slice slice : container.slices) {
```

Comment:
Being able to ignore the indexing parts when we don't need them is my motivation for making `IndexableSlice` inherit from `Slice`